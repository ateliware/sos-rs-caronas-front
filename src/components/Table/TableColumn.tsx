import { ReactNode } from 'react';

import TableButton from './TableButton';
import { Filters } from './Table';
import { FilterButton, SortButton } from './TableButtons';

type CellProps = {
  container?: boolean;
  ellipsis?: boolean;
  cellAlign?: 'start' | 'end' | 'center' | 'between';
  verticalAlign?: 'top' | 'middle' | 'bottom';
  textAlign?: 'left' | 'center' | 'right';
  render?: (value: any, rowData: any) => ReactNode;
};

export type ColumnProps = {
  fromKey: string;
  header: ReactNode;
  headerAlign?: 'start' | 'end' | 'center' | 'between';
  disableSorting?: boolean;
  filterType?: 'text' | 'number' | 'date' | 'select';
  filterOptions?: { label: string; value: string | boolean | number }[];
  onAction?: (key: string) => void;
  setFilters?: React.Dispatch<React.SetStateAction<Filters>>;
  filters?: Filters;
  width?: string;
  className?: string;
} & CellProps;

export default function TableColumn(props: ColumnProps) {
  const key = props.fromKey;
  const thClassName = ['table__cell'];
  const thCellContentClassName = ['table__cell__content'];

  if (props.className) {
    thClassName.push(props.className);
  }

  if (props.headerAlign) {
    thCellContentClassName.push(`justify-${props.headerAlign}`);
    thCellContentClassName.push(`text-${props.headerAlign}`);
  }

  return (
    <th style={{ width: props.width }} className={thClassName.join(' ')}>
      <div className={thCellContentClassName.join(' ')}>
        {props.header}
        {!props.disableSorting && <SortButton {...props} />}
        {props.filterType && <FilterButton {...props} />}
        {props.onAction && (
          <TableButton onClick={() => props.onAction?.(key)}>
            more_horiz
          </TableButton>
        )}
      </div>
    </th>
  );
}
