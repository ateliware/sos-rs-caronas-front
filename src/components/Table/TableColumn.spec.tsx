import { fireEvent, render } from '@testing-library/react';

import TableColumn from './TableColumn';

describe('TableColumn', () => {
  it('should render', () => {
    const { container } = render(
      <table>
        <thead>
          <tr>
            <TableColumn fromKey="name" header="Name" />
          </tr>
        </thead>
      </table>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render with variants', () => {
    const onAction = jest.fn();
    const { container } = render(
      <table>
        <thead>
          <tr>
            <TableColumn
              className="custom"
              headerAlign="center"
              fromKey="name"
              header="Name"
              width="100px"
              filterType="text"
              onAction={onAction}
            />
          </tr>
        </thead>
      </table>
    );
    expect(container).toMatchSnapshot();

    const th = container.querySelector('th');
    expect(th).toHaveClass('table__cell custom');
    expect(th).toHaveStyle('width: 100px');

    const buttons = container.querySelector('button');
    fireEvent.click(buttons as Element);
    expect(onAction).toBeCalledWith('name');
  });
});
