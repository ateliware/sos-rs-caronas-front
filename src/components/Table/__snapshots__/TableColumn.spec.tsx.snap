// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableColumn should render 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th
          class="table__cell"
        >
          <div
            class="table__cell__content"
          >
            Name
          </div>
        </th>
      </tr>
    </thead>
  </table>
</div>
`;

exports[`TableColumn should render with variants 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th
          class="table__cell custom"
          style="width: 100px;"
        >
          <div
            class="table__cell__content justify-center text-center"
          >
            Name
            <button
              class="btn btn--icon-square btn--table-action ml-s-100"
            >
              <span
                class="btn__body"
              >
                <span
                  class="btn__body__icon material-symbols-rounded"
                >
                  more_horiz
                </span>
              </span>
            </button>
          </div>
        </th>
      </tr>
    </thead>
  </table>
</div>
`;
